### DISCUSSION TITLE
External postcode requirement

### DISCUSSION THREAD

**Discussion by Cj-Malone** [osm-search/Nominatim#2348] [Geocoding around the world]
Hello, I was just wondering at what stage would a start recommending an external source of postcodes. Like CodePoint Open is in the docs, and (if you run nominatim.openstreetmap.org) when would extra postcodes be included there.

At the moment there are 13,282 published under the OGLv3 for Northern Ireland. Is this amount to insignificant?

**Comment by mtmail** [comment]
For CodePoint Open we have a conversion script https://github.com/osm-search/gb-postcode-data The next release of Nominatim has a new feature were it will look for one CSV file per country, preview of the documentation at http://nominatim.org/release-docs/develop/admin/Import/#external-postcodes

Not sure if Northern Ireland counts as its own country or not (GB vs UK).

Can you post the URL of the Northern Ireland data?

**Reply by Cj-Malone** [reply]
Technically Northern Ireland would be GB, although for it's usually easier to separate it out. Is the 1 file per county a hard limit?

The data comes from FHRS Open Data, I've split it out into a csv [here](https://github.com/osm-uk/NI-Postcodes). It'll be public and automated soon.

**Reply by lonvia** [reply]
Yes, that is a hard limit. Nominatim searches for exactly the one file per country. Catting together two CSV files should be fairly trivial though.

**Reply by Cj-Malone** [reply]
Yeah cating 2 files is nothing, if we (OSM UK) started publishing the extract of postcodes would you be interested in pulling it into the gb-postcode-data repo? Or would we could pull that data in and form one file on our end.

**Comment by lonvia** [comment]
We have to distinguish between Nominatim, the software, and nominatim.osm.org, the web service here.

I won't add any further external data sources for nominatim.openstreetmap.org. The instance is there for the benefit of mappers and any external data source hides the true state of OSM data.

Regarding Nominatim, the software, the goal is to make it easier to mix in such external sources. That has just happened for postcodes as @mtmail mentioned. But you need to run your own instance if you want to use that capability.

License-wise, if you want to use external postcodes, they have to be compatible with ODbL. Once you import them, they fall under the share-alike clause as they are mixed with OSM data.

**Reply by Cj-Malone** [reply]
Far enough. To be clear though, FHRS Open Data is available under the Open Government Licence and computable with ODbL, the same as CodePoint Open, it is actively used by mappers to get full addresses into OSM. Currently you basically can't search by post code in Northern Ireland because CodePoint Open excludes everything not England, Scotland and Wales, and existing OSM data isn't great there.

This answers the second question, even if we had all post codes in NI, you wouldn't want to include them in nominatim.openstreetmap.org.

What about including mentioning them in the docs for Nominatim?

**Reply by lonvia** [reply]
Adding links to free postcode sources in the docs would be fine. Somewhere in the 'external postcodes'  section for example.

In principle, we could even host the necessary conversion scripts under the osm-search orga, provided somebody else creates/maintains them. Maybe a generalised version of the `gb-postcode-data` repository.

**Reply by Cj-Malone** [reply]
Thank you, if we end up publishing postcodes I'll open a PR for the docs.

As for `gb-postcode-data` it looks like it is maintained by @mtmail, who works at @OpenCageData. @mtmail Will you be expanding that repo yourself to improve Open Cage, or would you like me to work on it?

**Reply by mtmail** [reply]
@Cj-Malone I sent you an email so we can figure out how we can work together on NI postcodes.
